packages/engine/Specs/Scene/Model/RuntimeModelInstancingPipelineStageSpec.js

```diff
@@ -0,0 +1,289 @@
+        const renderResources = mockRenderResources(node);
+        const runtimeNode = renderResources.runtimeNode;
+
+        scene.renderForSpecs();
+        RuntimeModelInstancingPipelineStage.process(
+          renderResources,
+          node,
+          scene.frameState,
+        );
+
+        expect(renderResources.attributes.length).toBe(5);
+
+        const shaderBuilder = renderResources.shaderBuilder;
+        ShaderBuilderTester.expectHasVertexDefines(shaderBuilder, [
+          "HAS_INSTANCING",
+          "USE_API_INSTANCING",
+        ]);
+        ShaderBuilderTester.expectHasFragmentDefines(shaderBuilder, [
+          "HAS_INSTANCING",
+        ]);
+        ShaderBuilderTester.expectHasAttributes(shaderBuilder, undefined, [
+          "in vec4 a_instancingTransformRow0;",
+          "in vec4 a_instancingTransformRow1;",
+          "in vec4 a_instancingTransformRow2;",
+          "in vec3 a_instancingPositionHigh;",
+          "in vec3 a_instancingPositionLow;",
+        ]);
+
+        ShaderBuilderTester.expectHasVertexUniforms(shaderBuilder, [
+          "uniform mat4 u_instance_nodeTransform;",
+        ]);
+
+        expect(runtimeNode.instancingTransformsBuffer).toBeDefined();
+        // The resource will be counted by NodeStatisticsPipelineStage.
+        expect(renderResources.model.statistics.geometryByteLength).toBe(0);
+      });
+    });
+
+    it("correctly creates transform matrices", function () {
+      return loadGltf(sampleGltfUrl).then(function (gltfLoader) {
+        const components = gltfLoader.components;
+        const node = components.nodes[0];
+        const renderResources = mockRenderResources(node);
+        const modelInstances = renderResources.model.sceneGraph.modelInstances;
+
+        const expectedTransformsTypedArray = new Float32Array([
+          -0.410076379776001, 0.7071067690849304, 0.576053261756897, 0,
+          -0.410076379776001, -0.7071067690849304, 0.576053261756897, 0,
+          0.8146623373031616, 0, 0.5799355506896973, 0, 0, 0, 0, 10, 10, 10,
+          -0.410076379776001, 0.7071067690849304, 0.576053261756897, 0,
+          -0.410076379776001, -0.7071067690849304, 0.576053261756897, 0,
+          0.8146623373031616, 0, 0.5799355506896973, 0, 0, 0, 0, 20, 20, 20,
+        ]);
+        const transformsTypedArray =
+          RuntimeModelInstancingPipelineStage._getTransformsTypedArray(
+            modelInstances,
+          );
+
+        expect(transformsTypedArray.length).toEqual(
+          expectedTransformsTypedArray.length,
+        );
+        for (let i = 0; i < expectedTransformsTypedArray.length; i++) {
+          expect(transformsTypedArray[i]).toEqualEpsilon(
+            expectedTransformsTypedArray[i],
+            CesiumMath.EPSILON10,
+          );
+        }
+      });
+    });
+
+    it("model instances update stage updates transform vertex attributes", function () {
+      return loadGltf(sampleGltfUrl).then(function (gltfLoader) {
```

Comment:
Let's prefer `async/await` syntax to promise chains. It flattens the test code and helps makes things more readable.